src/charm.py

```diff
@@ -1,104 +1,392 @@
+        cu_config = self._generate_cu_config()
+        if config_update_required := self._is_cu_config_up_to_date(cu_config):
+            self._write_config_file(content=cu_config)
+        service_restart_required = config_update_required
+        self._configure_pebble(restart=service_restart_required)
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
+        self._update_fiveg_f1_relation_data()
+        self._update_fiveg_gnb_identity_relation_data()
 
-        Learn more about actions at https://juju.is/docs/sdk/actions
+    def _relation_created(self, relation_name: str) -> bool:
+        """Return whether a given Juju relation was created.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: Whether the relation was created.
         """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
+        return bool(self.model.relations.get(relation_name))
+
+    def _generate_cu_config(self) -> str:
+        if self._f1_provider.requirer_f1_port:
+            du_f1_port = self._f1_provider.requirer_f1_port
         else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+            logger.info(
+                "DU F1 port information not available. Using default value %s", DU_F1_DEFAULT_PORT
+            )
+            du_f1_port = DU_F1_DEFAULT_PORT
+        return _render_config_file(
+            gnb_name=self._gnb_name,
+            cu_f1_interface_name=self._charm_config.f1_interface_name,
+            cu_f1_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            cu_f1_port=self._charm_config.f1_port,
+            du_f1_port=du_f1_port,
+            cu_n2_interface_name=self._charm_config.n2_interface_name,
+            cu_n2_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            cu_n3_interface_name=self._charm_config.n3_interface_name,
+            cu_n3_ip_address=_get_pod_ip(),  # type: ignore[arg-type]
+            amf_external_address=self._n2_requirer.amf_ip_address,
+            mcc=self._charm_config.mcc,
+            mnc=self._charm_config.mnc,
+            sst=self._charm_config.sst,
+            tac=self._charm_config.tac,
+        )
+
+    def _is_cu_config_up_to_date(self, content: str) -> bool:
+        """Decide whether config update is required by checking existence and config content.
+
+        Args:
+            content (str): desired config file content
+
+        Returns:
+            True if config update is required else False
+        """
+        if not self._config_file_is_written() or not self._config_file_content_matches(
+            content=content
+        ):
+            return True
+        return False
```

Comment:
```suggestion
    return not self._config_file_is_written() or not self._config_file_content_matches(content=content)
```

Reply:
Actually this whole method doesn't make any sense. I've refactored it completely.